### PR TITLE
Add track stitching class and example.

### PR DESCRIPTION
**Pull request by spike-dstl**
Co-authored-by: Scott Florence <sflorence@dstl.gov.uk>

Pull request to add track stitching example and 'Stitcher' class to main stone soup repository.

### CODE REVIEW

**Comment by sdhiscocks** [comment]
Code coverage could be improved. It looks like the *backwards* methods aren't tested currently.

File: stonesoup/tests/test_stitcher.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+from stonesoup.dataassociator.neighbour import GNNWith2DAssignment
+from stonesoup.deleter.error import CovarianceBasedDeleter
+from stonesoup.deleter.multi import CompositeDeleter
+from stonesoup.deleter.time import UpdateTimeStepsDeleter
+from stonesoup.initiator.simple import MultiMeasurementInitiator
+from stonesoup.types.groundtruth import GroundTruthPath, GroundTruthState
+from stonesoup.stitcher import TrackStitcher
+
+start_time = datetime.now().replace(second=0, microsecond=0)
+np.random.seed(100)
```
Comment (sdhiscocks) [inline stonesoup/tests/test_stitcher.py:22]: Can't really use seed in test (without resetting it again). Probably best to use seed in models, and also capture example starting points from this seed for now.

File: docs/examples/Track_Stitcher_Example.py [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+metric_manager.add_data(truths, set(all_tracks))
+
+plt.rcParams["figure.figsize"] = (10, 8)
+metrics = metric_manager.generate_metrics()
+
+siap_averages = {metrics.get(metric) for metric in metrics
+                 if metric.startswith("SIAP") and not metric.endswith(" at times")}
+siap_time_based = {metrics.get(metric) for metric in metrics if metric.endswith(' at times')}
+
+_ = SIAPTableGenerator(siap_averages).compute_metric()
```
Comment (sdhiscocks) [inline docs/examples/Track_Stitcher_Example.py:585]: Would be good to compare before and after, but I can see issue is that stitched tracks aren't interpolated. Probably should add that as separate component that could be used.

File: stonesoup/stitcher.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
+from datetime import timedelta
+from stonesoup.types.track import Track
+from stonesoup.types.detection import Detection
+from stonesoup.models.measurement.linear import LinearGaussian
+from copy import deepcopy
+from scipy.optimize import linear_sum_assignment
```
Comment (sdhiscocks) [inline stonesoup/stitcher.py:6]:
```suggestion
from collections import defaultdict
from copy import deepcopy
from datetime import timedelta

from scipy.optimize import linear_sum_assignment

from .base import Base, Property
from .hypothesiser import Hypothesiser
from .models.measurement.linear import LinearGaussian
from .types.detection import Detection
from .types.track import Track
```

File: docs/examples/Track_Stitcher_Example.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+        tracks = set(tracks)
+        for ids in combo:
+            x = []
+            for a in ids:
+                track = self.__get_track(a, tracks)
+                x = x + track.states
+                tracks.remove(track)
+            tracks.add(Track(x))
+
+        return tracks
```
Comment (sdhiscocks) [inline docs/examples/Track_Stitcher_Example.py:478]: Replace with just importing the new class.

**Review by sdhiscocks — CHANGES_REQUESTED** [review]

File: stonesoup/tests/test_stitcher.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                associated_measurements.add(hypothesis.measurement)
+            else:
+                track.append(hypothesis.prediction)
+        del_tracks = deleter.delete_tracks(tracks)
+        tracks -= del_tracks
+        tracks |= initiator.initiate(measurements - associated_measurements,
+                                     start_time + timedelta(seconds=n))
+        historic_tracks |= del_tracks
+    historic_tracks |= tracks
+    return historic_tracks
```
[the remark this answers is not in the document]
Comment (sdhiscocks) [reply stonesoup/tests/test_stitcher.py:72]: Could just use the standard `MultiTargetTracker`

**Comment by spike-dstl** [comment]
> Code coverage could be improved. It looks like the _backwards_ methods aren't tested currently.

File: stonesoup/stitcher.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    forward_hypothesiser: Hypothesiser = Property(
+        doc="Forward predicting hypothesiser.", default=None)
+    backward_hypothesiser: Hypothesiser = Property(
+        doc="Backward predicting hypothesiser.", default=None)
+    search_window: timedelta = Property(default=timedelta(seconds=30))
+
+    @staticmethod
+    @lru_cache()
+    def _extract_detection(track, backward=False):
+        state = track[0] if not backward else track[-1]
```
[the remark this answers is not in the document]
Comment (spike-dstl) [reply stonesoup/stitcher.py:36]: This gives a different result than wanted: 
The original method gives: 
backward = False -> state = track[0]
backward = True -> state = track[-1]

The new method would give: 
backward = False -> state = track[-1]
backward = True -> state = track[0]

I think your point is correct though, and the below change should be correct?
```suggestion
    def _extract_detection(track, backward=False):
        state = track[-1] if not backward else track[0]
```

**Review by sdhiscocks — APPROVED** [review]

**Pull request merged**